geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/cache/ThreadLocalCache.scala

```diff
@@ -0,0 +1,146 @@
+/***********************************************************************
+ * Copyright (c) 2013-2020 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.utils.cache
+
+import java.io.Closeable
+import java.lang.ref.WeakReference
+import java.util.concurrent._
+
+import com.github.benmanes.caffeine.cache.{Cache, Caffeine}
+import com.google.common.util.concurrent.MoreExecutors
+
+import scala.concurrent.duration.Duration
+
+/**
+ * Creates a per-thread cache of values with a timed expiration.
+ *
+ * Map operations will only affect/reflect the state of the current thread. Additional methods `globalIterator`
+ * and `estimatedGlobalSize` are provided for global views across all threads, which generally should only be used
+ * for debugging.
+ *
+ * Since caches are only cleaned up when accessed, uses an asynchronous thread to
+ * actively clean up any orphaned thread local values
+ *
+ * @tparam K key type
+ * @tparam V value type
+ */
+class ThreadLocalCache[K <: AnyRef, V <: AnyRef](
+    expiry: Duration,
+    executor: ScheduledExecutorService = ThreadLocalCache.executor
+  ) extends scala.collection.mutable.Map[K, V]
+    with scala.collection.mutable.MapLike[K, V, ThreadLocalCache[K, V]]
+    with Runnable
+    with Closeable {
+
+  import scala.collection.JavaConverters._
+
+  // weak references to our current caches, to allow cleanup + GC
+  private val references = new ConcurrentLinkedQueue[(Long, WeakReference[Cache[K, V]])]()
+
+  private val caches = new ThreadLocal[Cache[K, V]]() {
+    override def initialValue(): Cache[K, V] = {
+      val cache = Caffeine.newBuilder().expireAfterAccess(expiry.toMillis, TimeUnit.MILLISECONDS).build[K, V]()
+      // this will always succeed as our queue is unbounded
+      references.offer((Thread.currentThread().getId, new WeakReference(cache)))
+      cache
```

Comment:
Ok, how does having a WeakReference to the 'cache' help?  

It seems like each thread will have a non-Weak reference to the same object, right?

Reply:
yes for long-running threads, but if the calling thread dies then the cache could be gc'd and we don't want to prevent that.